blocks/algorithms.py

```diff
@@ -0,0 +1,227 @@
+"""Training algorithms."""
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+
+import theano
+from theano import tensor
+
+from blocks.utils import update_instance
+from blocks.graph import ComputationGraph
+
+
+class TrainingAlgorithm(object):
+    """Base class for training algorithms.
+
+    A training algorithm object has a simple life-cycle.
+    First it is initialized by calling its :meth:`initialize` method.
+    At this stage, for instance, Theano functions can be compiled.
+    After that the :meth:`process_batch` method is repeatedly
+    called with a batch of training data as a parameter.
+
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def initialize(self):
+        """Initialize the training algorithm."""
+        pass
+
+    @abstractmethod
+    def process_batch(self, batch):
+        """Process a batch of training data.
+
+        Attributes
+        ----------
+        batch : dict
+            A dictionary of (source name, data) pairs.
+
+        """
+        pass
+
+
+class TheanoTrainingAlgorithm(TrainingAlgorithm):
+    """A training algorithm based on Theano functions.
+
+    Parameters
+    ----------
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Attributes
+    ----------
+    updates : list of (shared variable, Theano expression) tuples
+        Updates to be done for every batch. It is required that the
+        updates are done using the old values of optimized parameters.
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Notes
+    -----
+        Changing `updates` attribute or calling `add_updates` after
+        the `initialize` method is called will have no effect.
+
+    .. todo::
+
+       Ideally the parameters should be extracted from the `cost`
+       computation graph. But in order for this to work our graph
+       traversal routine should recurse into ScanOp's, which is currently
+       not the case.
```

Comment:
On second thought. I think the best way to go is to keep it this way: Keep the parameters as an argument! This makes the training algorithm truly modular i.e. you can use it without the need to annotate the graph in one way or the other. I think it'd be great if people can actually build their own model entirely independently, and then just use a training algorithm class to get the Theano updates they need to perform. This is very much in line with the star-based dependency graph that we talked about.